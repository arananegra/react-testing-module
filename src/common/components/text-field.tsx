import * as React from 'react';
import MaterialTextField, { TextFieldProps as MaterialTextFieldProps } from '@material-ui/core/TextField';
//Removed due to tests conflicts
//import { Overwrite } from '@material-ui/types';

export interface TextFieldProps {
  name: string;
  label: string;
  value?: string;
  onChange: (field: string, value, event?) => void;
  error?: string;
  type?: string;
  onBlur?: (field: string, value, event?) => void;
  className?: string;
}

export const TextField: React.FC<TextFieldProps> = props => {
  const {
    name,
    label,
    onChange,
    value,
    error,
    type,
    onBlur,
    className,
    ...otherProps
  } = props;
  const hasError = Boolean(error);

  const handleChange = e => {
    onChange(name, e.target.value, e);
  };

  const handleBlur = e => {
    if (onBlur) {
      onBlur(name, e.target.value, e);
    }
  };

  return (
    <MaterialTextField
      {...otherProps}
      className={className}
      label={label}
      margin="normal"
      value={value}
      type={type}
      onChange={handleChange}
      inputProps={
        { ['data-testid']: 'mui-textfield' }
      }
      onBlur={handleBlur}
      error={hasError}
      helperText={hasError && error}
    />
  );
};

TextField.defaultProps = {
  type: 'text',
};
