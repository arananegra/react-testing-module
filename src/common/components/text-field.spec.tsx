import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { TextField } from './text-field';
import { BaseTextFieldProps, TextFieldProps } from '@material-ui/core/TextField';

describe('Button tests', () => {
  it('should render a happy path input with simple props', () => {
    // Arrange
    const props = {
      label: 'text field',
      name: 'input-name',
      onChange: jest.fn(),
      value: 'value'
    };

    // Act
    const { getByTestId, asFragment } = render(<TextField {...props} />);
    const textFieldElement = getByTestId('mui-textfield') as HTMLInputElement;
    //Assert

    expect(textFieldElement).not.toBeUndefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call on onChange when event is fired', () => {
    // Arrange
    const props = {
      label: 'text field',
      name: 'input-name',
      onChange: jest.fn(),
    };
    // Act
    const { getByTestId } = render(<TextField {...props} />);
    const textFieldElement = getByTestId('mui-textfield') as HTMLInputElement;

    //Assert
    expect(textFieldElement.value).toBe('');
    fireEvent.change(textFieldElement, { target: { value: 'John' } });
    expect(textFieldElement.value).toBe('John');

    expect(props.onChange).toHaveBeenCalled();
  });

  it('should call onBlur when event is passed', () => {
    // Arrange
    const props = {
      label: 'text field',
      name: 'input-name',
      value: '',
      onBlur: jest.fn(),
      onChange: jest.fn(),
    };

    // Act
    const { getByTestId } = render(<TextField {...props} />);
    const textFieldElement = getByTestId('mui-textfield') as HTMLInputElement;
    fireEvent.blur(textFieldElement);
    //Assert
    expect(props.onBlur).toHaveBeenCalled();
  });

});
