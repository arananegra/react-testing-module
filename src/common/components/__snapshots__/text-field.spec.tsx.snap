// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests should render a happy path input with simple props 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
    data-testid="external-mui-input"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
    >
      text field
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        data-testid="mui-textfield"
        type="text"
        value="value"
      />
    </div>
  </div>
</DocumentFragment>
`;
